src/NumericalAlgorithms/Spectral/SwshCollocation.hpp

```diff
@@ -0,0 +1,245 @@
+  /// `precomputed_spherical_harmonic_collocation` function
+  explicit Collocation(size_t l_max) noexcept;
+
+  /// default constructor required for iterator use
+  ~Collocation() = default;
+  Collocation() = default;
+  Collocation(const Collocation&) = default;
+  Collocation(Collocation&&) = default;
+  Collocation& operator=(Collocation&) = default;
+  Collocation& operator=(Collocation&&) = default;
+
+  /// retrieve the `sharp_geom_info*` stored. This should largely be used only
+  /// for passing to other libsharp functions. Otherwise, access elements
+  /// through iterator or access functions.
+  // clang-tidy seems to think the pointer will implicitly convert to the
+  // libsharp type. It does not.
+  sharp_geom_info* get_sharp_geom_info() const noexcept {
+    return *(geom_info_.get());  // NOLINT
+  }
+
+  /// Retrieve the \f$\theta\f$ value for a given index in a libsharp-compatible
+  /// array
+  double theta(size_t offset) const noexcept;
+  /// Retrieve the \f$\phi\f$ value for a given index in a libsharp-compatible
+  /// array
+  double phi(size_t offset) const noexcept;
+
+  constexpr size_t l_max() const noexcept { return l_max_; }
+
+  /// Compute the number of entries the libsharp-compatible data structure
+  /// should have
+  constexpr size_t size() const noexcept {
+    return (l_max_ + 1) * (2 * l_max_ + 1);
+  }
+
+  // @{
+  /// Get a bidirectional iterator to the start of the grid. `operator*` for
+  /// that iterator gives a
+  /// `std::pair<size_t /*offset*/ std::pair<double /*theta*/, double /*phi*/>`
+  const Collocation<Representation>::CollocationIterator begin() const
+      noexcept {
+    return CollocationIterator{make_not_null(this), 0};
+  }
+  // @}
+  // @{
+  /// Get a bidirectional iterator to the end of the grid. `operator*` for
+  /// that iterator gives a
+  /// `std::pair<size_t /*offset*/ std::pair<double /*theta*/, double /*phi*/>`
+  const Collocation<Representation>::CollocationIterator end() const noexcept {
+    return CollocationIterator{make_not_null(this), size()};
+  };
+  // @}
+
+ private:
+  // an extra pointer layer is required for the peculiar way that libsharp
+  // constructs these values.
+  size_t l_max_ = 0;
+  std::unique_ptr<sharp_geom_info*, detail::DestroySharpGeometry> geom_info_;
+};
+
+namespace detail {
+// We use a `std::index_sequence` to loop over the functions
+// `precomputed_static_collocation_impl` we would like to only compute and store
+// a specific `l_max` that is requested, not all values up to
+// `SWSH_COLLOCATION_MAXIMUM_L_MAX`. This cannot be done in the more traditional
+// way of either having a `std::vector` or `std::undordered_map` that is
+// `static` because only the construction phase is guaranteed to be thread-safe,
+// not the access. This would mean were we to use a `std::vector` or
+// `std::unordered_map` we would need to compute all values up to
+// `MAXIMUM_L_MAX` immediately. By doing the loop over `Is` we are able to
```

Comment:
`MAXIMUM_L_MAX` -> `SWSH_COLLOCATION_MAXIMUM_L_MAX` (sorry, I should've had this in my last comment already)